File: day5.py
with open('input/day5', 'r') as f:
    lines = [i.replace('\n', '') for i in f.readlines()]

stacks1: list[list] = []

for i, line in enumerate(lines):
    line = line.strip()
    if line[0].isdigit():
        [stacks1.append([]) for _ in range(int(line.split(' ')[-1]))]
        for line2 in lines[i-1::-1]:
            [stacks1[j].append(k.strip('[]')) for j, k in enumerate(line2.replace('    ', ' ').split(' ')) if k != '']
        break

stacks2 = [x[:] for x in stacks1]

for line in lines[i+2:]:    # PART 1
    how_much, origin, dest = [int(i) for i in [j for j in line.split(' ') if j.isdigit()]]
    [stacks1[dest-1].append(stacks1[origin-1].pop()) for _ in range(how_much)]

    stacks2[dest-1].extend(stacks2[origin-1][len(stacks2[origin-1])-how_much:])
    del stacks2[origin-1][len(stacks2[origin-1])-how_much:]

print('1) Last crates', ''.join([i[-1] for i in stacks1]))
print('2) Last crates', ''.join([i[-1] for i in stacks2]))

File: day5.js
let fs = require("fs");
const { join } = require("path");

fs.readFile("input/day5", "utf8", (err, data) => {
    const lines = data.trim().split('\n').map(i => i.trim());
    const stacks = [];
    let i;
    for (i = 0; i < lines.length; i++) {
        const line = lines[i].trim();
        if (/^\d+$/.test(line[0])) {
            line.split(' ').forEach(j => (j != '') && (stacks.push([])));

            for (let i2 = i - 1; i2 >= 0; i2--) {
                lines[i2].replaceAll('    ', ' ').split(' ').forEach((j, k) => (j != '') && (stacks[k].push(j.replace('[', '').replace(']', ''))));
            }
            break;
        }
    }

    const stacks2 = stacks.map(j => [...j]);
    // console.log(stacks);
    lines.slice(i + 2).forEach((line => {
        let [how_much, origin, dest] = line.split(' ').filter(j => /^\d+$/.test(j)).map(j => parseInt(j));
        [...Array(how_much)].forEach(_ => stacks[dest - 1].push(stacks[origin - 1].pop()));

        stacks2[dest - 1].push(...stacks2[origin - 1].slice(stacks2[origin - 1].length - how_much));
        stacks2[origin - 1] = stacks2[origin - 1].slice(0, stacks2[origin - 1].length - how_much);
    }));
    console.log('1) Last crates', stacks.map(j => j[j.length - 1]).join(''));
    console.log('2) Last crates', stacks2.map(j => j[j.length - 1]).join(''));
});
